src/invariant/LedgerEntryIsValid.cpp

```diff
@@ -0,0 +1,225 @@
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "invariant/InvariantManager.h"
+#include "invariant/LedgerEntryIsValid.h"
+#include "ledger/LedgerDelta.h"
+#include "lib/util/format.h"
+#include "main/Application.h"
+#include "xdrpp/printer.h"
+
+namespace stellar
+{
+
+std::shared_ptr<Invariant>
+LedgerEntryIsValid::registerInvariant(Application& app)
+{
+    return app.getInvariantManager().registerInvariant<LedgerEntryIsValid>();
+}
+
+std::string
+LedgerEntryIsValid::getName() const
+{
+    return "LedgerEntryIsValid";
+}
+
+bool
+LedgerEntryIsValid::isStrict() const
+{
+    return false;
+}
+
+std::string
+LedgerEntryIsValid::checkOnOperationApply(Operation const& operation,
+                                          OperationResult const& result,
+                                          LedgerDelta const& delta)
+{
+    uint32_t currLedgerSeq = delta.getHeader().ledgerSeq;
+    if (currLedgerSeq > INT32_MAX)
+    {
+        return fmt::format("LedgerHeader ledgerSeq ({}) exceeds limits ({})",
+                           currLedgerSeq, INT32_MAX);
+    }
+
+    std::string msg = check(delta.addedBegin(), delta.addedEnd(),
+                            currLedgerSeq);
+    if (!msg.empty())
+    {
+        return msg;
+    }
+
+    msg = check(delta.modifiedBegin(), delta.modifiedEnd(),
+                currLedgerSeq);
+    if (!msg.empty())
+    {
+        return msg;
+    }
+    return {};
+}
+
+template <typename IterType>
+std::string
+LedgerEntryIsValid::check(IterType iter, IterType const& end,
+                          uint32_t ledgerSeq) const
+{
+    for ( ; iter != end; ++iter)
+    {
+        auto s = checkIsValid(iter->current->mEntry, ledgerSeq);
+        if (!s.empty())
+        {
+            s += ": ";
+            s += xdr::xdr_to_string(iter->current->mEntry);
+            return s;
+        }
+    }
+    return {};
+}
+
+std::string
+LedgerEntryIsValid::checkIsValid(LedgerEntry const& le,
+                                 uint32_t ledgerSeq) const
+{
+    if (le.lastModifiedLedgerSeq != ledgerSeq)
+    {
+        return fmt::format("LedgerEntry lastModifiedLedgerSeq ({}) must equal"
+                           " LedgerHeader ledgerSeq ({})",
+                           le.lastModifiedLedgerSeq, ledgerSeq);
+    }
+    switch (le.data.type())
+    {
+    case ACCOUNT:
+        return checkIsValid(le.data.account());
+    case TRUSTLINE:
+        return checkIsValid(le.data.trustLine());
+    case OFFER:
+        return checkIsValid(le.data.offer());
+    case DATA:
+        return checkIsValid(le.data.data());
+    default:
+        return "LedgerEntry has invalid type";
+    }
+}
+
+std::string
+LedgerEntryIsValid::checkIsValid(AccountEntry const& ae) const
+{
+    if (ae.balance < 0)
+    {
+        return fmt::format("Account balance ({}) cannot be negative",
```

Comment:
It should say "Account balance ({}) is negative", as it is listing errors here, not conditions, so the wording is consistent.